src/levanter/data/shardcache.py

```diff
@@ -0,0 +1,712 @@
+#   (This metadata can be used for seeking.)
+
+
+# How do we build the cache?
+# We assign each writer to a subset of the input shards. Each writer then builds one chunk from each of its shards
+# in round robin order. Whenever a chunk is finished, we write it and its metadata to disk. We also update
+# the metadata for the input shard and notify the readers that a new chunk is available.
+
+# What is the global ordering of data for sharded reads? (for training data)
+# We pretend the list of data is infinite by cycling.
+# Given a list of chunks and R*, we define the global ordering over data as follows:
+# First define R* iterators over chunks, with chunk_iterator[r] being defined as flatten_chunks(loop(all_chunks)[r::R*])
+# Then the global ordering is defined as:
+# chunk_iterator[0][0], chunk_iterator[1][0], chunk_iterator[2][0], ..., chunk_iterator[R*-1][0],
+# chunk_iterator[0][1], chunk_iterator[1][1], ..., chunk_iterator[R*-1][1], ...
+# that is, example[i] == chunk_iterator[i % R*][i // R*]
+
+# if we have R* readers, then each reader_iterator[r][j] == chunk_iterator[r][j]
+# Moreover, if either R or R* is a multiple of the other, then we still get a nice property where
+# each reader reads from a strided slice of the chunk_iterators:
+
+# (Boring math)
+# If we have R readers, then reader_iterator[r][j] == example[j * R + r] == chunk_iterator[(j * R + r) % R*][(j * R + r) // R*]
+# If we have R == n * R*, then reader_iterator[r][j] == example[j * R + r] == chunk_iterator[(j * R + r) % R*][(j * R + r) // R*]
+#  == chunk_iterator[r % R*][(j * n * R* + r) // R*] == chunk_iterator[r % R*][j * n + r // R*], so each reader reads from
+# a strided slice (specifically islice(..., r//R*, None, n))
+# If we have R* == n * R, then reader_iterator[r][j] == example[j * R + r] == chunk_iterator[(j * R + r) % R*][(j * R + r) // R*]
+# == chunk_iterator[R * (j % n) + r][(j * R + r) // R*] and so each reader reads from n different chunk_iterators.
+# so we round-robin over a slice of the chunk_iterators.
+
+# For other cases (R and R* don't divide each other), there's no simple relationship between the reader and chunk iterators
+# and you end up reading from everywhere, but that's ok.
+
+# What is the global ordering of data for epochal, unsharded reads? (for validation data)
+# When we want to do a single pass over the data, we don't cycle and we don't shuffle. We just read the data in order.
+
+# Coordination
+# We use Ray to coordinate writing. We instantiate one ray task per writer (typically one per machine), and each task
+# is given a list of shards to process. Each task then processes one chunk from each shard in round robin order.
+# When a chunk is finished, we write it to disk and notify that a new chunk is available. There is a single ray actor
+# that is responsible for handling which chunks are available and which are not. This actor exposes a method
+# to get a chunk by index
```

Comment:
this is what i was going for. I was starting to move to one task per shard and even each shard-task spawning off tasks per "chunk" (which would allow more parallelism after moving to defining chunk size in terms of number of rows)